### ISSUE TITLE
adding new notification feature.

### ISSUE THREAD

**Issue by sagarkaurav** [franciscop/picnic#149]
I'm working on new notification feature. So we can display short notification message on user's viewport.Here is some code that i have been working on.
[https://jsfiddle.net/sagar313/n7omj3aL/](url)

**Comment by franciscop** [comment]
That looks quite good! It is the kind of plugin that would be lovely to have some sort of official system to include it. Unfortunately I never made a system to accept/showcase external plugins, how do you suggest we include it? I think it has a too narrow use to include it for every developer, but it's really nice and would love to make it somewhat official.

Would you be willing to create a "Plugins" section in the website? Not too difficult, but not trivial and I am unable since my focus for active development has [shifted to something else](https://serverjs.io/). It could consist of a button in the `<nav>` with the puzzle icon called "Plugins" for example. Let me know if you might be interested so I can help you out setting things up.

**Comment by sagarkaurav** [comment]
That would be great :smile: . It will also minimize the main css file because developer only include plugin code that they really need in their particular project .

**Comment by franciscop** [comment]
Oh yeah, but the main benefit I am looking for is for people to be able to add any kind of feature they want freely, since I have been rejecting over the years some great propositions that didn't set right with the core library.

So, roughly speaking we'd need to:

1. Add a new empty folder `plugins` in the root.
2. Create all the small front-end changes. Maybe copy/paste the current `/documentation`.
3. Modify the Gruntfile.js and add some scripting to compile each of the plugins automatically.

That is for the website, for the plugin system I think it'd be possible to do:

- Make it so **the processed CSS** can be included from a public path such as `https://unpkg.com/picnic@6/plugins/notifications/notifications.css` (not existing yet). JS also or just provide a snippet if it's small.
- Make it so it can be included **from the SCSS** from an internal path such as `@import "./node_modules/picnic/plugins/notification/notification";`

I think that to achieve this, the plugin must be using the default theme (which means it will follow any variable set in SCSS) and any direct dependency with the abstract `_class.scss`. Then the Grunt will compile each of the plugins on build time and everything can be uploaded.

Sounds good? What parts can you do? What parts you don't want to do?

**Comment by sagarkaurav** [comment]
I will manage first two points creating plugins folder and chnaging front end. I think we can create new branch to work on these changes instead of pushing all changes directly to master.

**Comment by franciscop** [comment]
As you prefer, I don't mind using any branch since changes must be manually tagged and published to be released to other users, CDNs and all. I will get some time on the weekend, I'll try to use it on modifying the Gruntfile.

**Comment by sagarkaurav** [comment]
Here's what what i am thinking first we create new branch then make our changes then merge it to master branch make it release with major version. What you think?

**Comment by sagarkaurav** [comment]
@franciscop 
Here is my suggestion to make library more flexible.

1. First we will create ` _variable.scss` file which will hold all the core variables used by picnic css. And that file will act as default theme because right now`  src/themes/default/*`  directory only declaring the variable. If any plugin wants to use core variables they can include this file easily.

2. Rename `src/plugin` to `src/core` by doing that it will make other people clear that the code inside that directory are making core features. And all the plugin inside it chould import `_variable.scss` if they need some core variable's value.

3. Create new file called `_picnic_core.scss` will import all the core plugings.
```scss
@import './vendor/compass-breakpoint/stylesheets/breakpoint';
//impoting core plugins
@import './core/normalize/normalize';
@import './core/generic/generic';
@import './core/fontello/fontello';
@import './core/nav/nav';
//all the others core import will go here.
```

4. The `src/themes` directory  where all other themes will go. Here theme can import core variables .i.e., `_variable.scss` and overright and also can add new scss. The developer can include theme in there `scss` file just by importing theme in `src/picnic.scss` file.

```scss
@import 'variable.scss';
@import 'themes/the_theme/the_theme.scss'
@import 'picnic_core';
```

If we dont want include any theme we can go with default.
```scss
@import 'variable.scss';
@import 'picnic_core';
```

Note: Why we need these changes right now the tabs plugin is using `$picnic-separation` variable in it which is defined inside default theme if we want to move it to external plugin i.e., (inside `/plugins` directory) it will start througing error because it can't find that variable. But if we use above solution we can solve this problem just by including `_variable.scss` file inside tabs plugin's `_class.scss` file. It also help us making new plugin's if they need some basic variables. 

>But changing core vraiable's value will not be allowed inside plugin files.

**Comment by franciscop** [comment]
1. Sure, this file would be in the root, right? Actually, I have never used themes and from the state of it I don't think anyone has ever used them, so it'd be totally okay to just remove them. What I do now (and I think it's a good way) if I want to modify the default theme is to edit the variables before including picnic.scss, [see here](https://github.com/franciscop/server/blob/master/docs/assets/style.scss).

2. Agree as well.

3. Do not agree. The library went down this route few major versions back and it was a real mess TBH. There should be no doubt of what file is Picnic and how to include it (and only one). **Then** there are plugins as well, but the main/core library name is and should be kept **picnic.scss**. I would even go so far as to say that `picnic.scss` should be in the root to make it easier, but that I don't mind so much.

4. This wouldn't really work (but see 1.), since `_variables.scss` would set the variables and so the theme cannot set them because of the `!default` (which makes a variable get set it only if it's undefined, they are defined in `_variables.scss`). We also cannot remove the flag from a theme nor from the variables since all the options should be user-overridable. Since we will be releasing a major version I suggest to just remove the themes and leave a `_variables.scss` on the root.

**Comment by sagarkaurav** [comment]
The conclusion of our discussion is that 
1.  Themes will be completely removed.
2.  `_variables.scss` file in root directory will contain all the core variables
3.  the following directory structure we will follow:
![dir](https://user-images.githubusercontent.com/8871080/27396145-d40a1ca0-56cf-11e7-89d4-fde9fd007d79.png)



### Tell me if i am missing something. If its looks good notify me so i start moving things.

**Comment by franciscop** [comment]
That sounds great :+1:

**Comment by franciscop** [comment]
Hey @sagarkaurav could you do something else? Otherwise I'll try to move it forward myself.

**Comment by sagarkaurav** [comment]
@franciscop I was pretty busy recently. I will try to work over the weekend.  I will try to update jade to pug this weekend.

**Comment by sagarkaurav** [comment]
@franciscop please review the changes made in develop branch.

**Comment by franciscop** [comment]
@sagarkaurav thanks, will do ASAP :smile: (weekend+)

**Comment by sagarkaurav** [comment]
@franciscop have you reviewed code yet.

**Comment by franciscop** [comment]
Yes I did an overview of it, now I am looking at it better though, sorry for the delay. Couple of notes:

## Side nav [bug]

The side nav buttons seem to no longer have `href`s. They came from JS, which in turn came from the `id` in each `<h2>`. They no longer have those ids being autogenerated, it seems to be because now we're using `include:markdown-it` instead of `include:markdown` (from `marked`).

I don't mind the markdown engine at all, but if we're switching it then this would have to be fixed (I'm curious though about why changing the markdown engine).


## Normalize [bug]

We have to make more explicit how to include the plugins. The `/test` page shows the buttons and table broken; this is because the plugin `tabs` has the full `normalize`, which is included after the main library, so it overwrites part of the core styles.

I think it makes sense that the plugins are an extra of the main Picnic CSS, so it is required that Picnic is included before and then the plugin (then you don't need to import normalize on each plugin).

```js
<link rel='stylesheet' href='https://unpkg.com/picnic@6/picnic.min.css'>
<link rel='stylesheet' href='https://unpkg.com/picnic@6/plugins/notification/plugin.min.css'>
```

It might also make sense to want to make the plugin a standalone feature, but this would be an extra, the main feature of plugins being part of picnic as above. As a standalone it could be:

```js
<link rel='stylesheet' href='https://unpkg.com/picnic@6.4/plugins/notification/standalone.min.css'>
```

> Note: `plugin.min.css` and `standalone.min.css` are temporary placeholders, use whichever you think are better if following this route.


## CTA button [opinion]

I would leave only one CTA button in the nav, which I am not 100% sure whether it'd be `Documentation` (of the internal plugins) or `Plugins`, but I am guessing the latter makes more sense since we want people contributing. So I'd make `Documentation` match the other pseudo-buttons but put it as the first button to make sure it is not missed.


## Conclusion

There are few things like examples in the documentation, etc but I think those are not a priority right now and they are more of an on-going basis. Everything else looks great and ready to go, thank you a lot for doing this work to migrate it! Please make sure to add your name anywhere relevant in Github or the website.

**Comment by sagarkaurav** [comment]
@franciscop are you still interested in the development of next version.
